### PR TITLE
Refactored to take new memory model into account

### PR DESCRIPTION
**Pull request by fgmccabe**
Also expanded and rewritten explanations of memory management

### CODE REVIEW

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 1]
```diff
@@ -133,17 +132,17 @@ it:
...
-  finally
-    call $free
+  call-export "ex:malloc"
+  own (i32)
+    call-export "ex:free"
+  end
+  let (result u32) (local $tgt (owned i32))
+    local.get $str
+    local.get $tgt
+	owned.access
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:142]: Weird indentation here

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 2]
```diff
@@ -160,30 +159,101 @@ There is a lot going on here, so we will take this slowly:
...
+   WebAssembly module.
 1. Since we want to avoid a memory leak, we want to arrange for the allocated
    memory to be freed. We arrange for this to occur at the end of the adapter
-   function by using a `deferred` instruction block -- whose `finally` sub-block
-   is executed when the memory can no longer be accessed.
-   
->   In effect, the address returned by `$malloc` is preserved by the `deferred`
-   block and revived when executing the `finally` sub-block.
+   function by using an `own` instruction block -- which when the memory can no
+   longet be access invokes its sub-block -- which releases the allocated memory
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:173]: Typos: s/longet/longer/, s/access/accessed/

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 3]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
-      finally
-        call $free
+      call-export "ex:malloc"
+      own (i32)
+        call-export "ex:free"
+      end
+      let (result u32) (local $tgt (owned i32))
+        local.get $str
+        local.get $tgt
+	    owned.access
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:315]: Weird indentation again; perhaps it's intentional for `owned.access`?

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 4]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
       local.get $str
-      local.get $tgt
-      string-to-memory "memx"
-      call $countCodes_impl
-      i32-to-u32
+      string.size
+      call-export "ex:malloc"
+	  local.tee $m1
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:347]: Indentation

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 5]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
+to release the memory allocated for the string.
+
+One straightfoward technique for this is to add additional local variables to
+the adapter function -- that are used to hold the references collected by the
+`own` instruction itself. 
 
-In the case of the `defer-scope` and `deferred` instructions we eliminate both
-and move the instructions within the `finally` sub-block to the end of the
-`defer-scope` block:
+At the same time, we can unwrap the various uses of `owned.access`, and invoke
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:333]: I found the use of "unwrap" here confusing...do you mean something like "desugar" (I realize that's its own term of art, will think of other ideas in the meantime)?

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 6]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
-  memory-to-string "memi"
-    
-  let (local $str string)
-    local.get $str
-    string.size
-    call $malloc
-    local.tee $def_0
-    let (local $tgt i32)
+  (param $ptr i32)(param $len i32) (result i32)
+  (local $m1 i32)
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:339]: Might make it clearer if you added a comment here, down at the tee, and at the call of free to make it clear what happened to the `own` (assuming I'm understanding it correctly).

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 1 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:142]: caused by emacs. Do not always remember to set untabify :(

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 4 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:347]: emacs

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 6 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:339]: will do

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 5 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:333]: Well, conceptually an owned value is encapsulated -- at least as far as types are concerned.
This is for verify-ability of the code.

Comment (ajklein) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:333]: reading this again with the code, I think I now understand better what you meant.

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 7]
```diff
@@ -77,36 +77,35 @@ that highlights that this is not a regular webAssembly function:
...
+  local.get $ptr
+  local.get $len
+  string.from.utf8 "memi"
+  call-import "countCodes"
+  i32.from.u32
 )
 ```
 
-The heart of this adapter is the special adapter instruction `memory-to-string`
+The heart of this adapter is the special adapter instruction `string.from.utf8`
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:88]: In Instructions.md, we have `string.lift_memory` and `string.lower_memory`.  Is `string.from.utf8` and `utf8.from.string` proposed anywhere?

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 8]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
       local.get $str
-      local.get $tgt
-      string-to-memory "memx"
-      call $countCodes_impl
-      i32-to-u32
+      string.size
+      call-export "ex:malloc"
+      local.tee $m1   ;; store captured references for later
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:347]: Thus far, we've avoided any `local.set`s or mutation of locals to ensure that adapter fusion can easily propagate defs to uses.

Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:347]: I understand. But, setting is unavoidable here.
Also, remember that the tee instruction is a result of the fusion. It is not present in the original adapters.

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 7 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:88]: Well, I am not much in favor of mentioning lower & lift in the instruction names.
I believe instruction names should state what they do, not their purpose. (This is especially true for low-level instruction names that could easily be repurposed.)
Also, I added utf8 as a signal that we may be supporting other encodings. This is where the encoding makes a difference.

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 9]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
+      let (result (u32) (local $tgt i32)
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:348]: Remove extra `(` before u32

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 10]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
 (func $countCodes
-  (param $ptr i32)(param $len i32) (result i32))
-  defer-scope
+  (param $ptr i32)(param $len i32) (result i32)
     local.get $ptr
     local.get $len
-    memory-to-string "memi"
-    let (local $str string)
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:305]: Should be `(result u32)`

File: proposals/interface-types/working-notes/scenarios/unicode_count.md [hunk 11]
```diff
@@ -223,100 +293,118 @@ module side-effecting resources in another.
...
-    string.size
-    call $malloc
-    local.tee $def_0
-    let (local $tgt i32)
+  (param $ptr i32)(param $len i32) (result i32)
+  (local $m1 i32)  ;; Used for owned block
+    local.get $ptr
+    local.get $len
+    string.from.utf8 "memi"
+    let (result i32)(local $str string)
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/unicode_count.md:343]: `(result u32)` here too

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 9 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:348]: got it

File: proposals/interface-types/working-notes/scenarios/unicode_count.md (hunk 8 above)
Comment (lukewagner) [reply proposals/interface-types/working-notes/scenarios/unicode_count.md:347]: Oh hah, sorry, I hadn't noticed this was only in the fused result.  Nevermind :)

**Pull request merged**